README
NAME

    Zing::Store::Rocks - RocksDB Storage

ABSTRACT

    RocksDB Storage Abstraction

SYNOPSIS

      use File::Spec;
      use Zing::Encoder::Dump;
      use Zing::Store::Rocks;
    
      my $store = Zing::Store::Rocks->new(
        root => File::Spec->catfile(
          File::Spec->tmpdir, rand
        ),
      );
    
      # $store->drop;

DESCRIPTION

    This package provides a RockDB-specific storage adapter for use with
    data persistence abstractions. The "client" attribute accepts a RocksDB
    object.

INHERITS

    This package inherits behaviors from:

    Zing::Store

LIBRARIES

    This package uses type constraints from:

    Zing::Types

ATTRIBUTES

    This package has the following attributes:

 client

      client(InstanceOf["RocksDB"])

    This attribute is read-only, accepts (InstanceOf["RocksDB"]) values,
    and is optional.

 root

      root(Str)

    This attribute is read-only, accepts (Str) values, and is optional.

METHODS

    This package implements the following methods:

 decode

      decode(Str $data) : HashRef

    The decode method decodes the data provided and returns the data as a
    hashref.

    decode example #1

        # given: synopsis
      
        $store->decode('{"status"=>"ok"}');

 drop

      drop(Str $key) : Int

    The drop method removes (drops) the item from the datastore.

    drop example #1

        # given: synopsis
      
        $store->drop('zing:main:global:model:temp');

 encode

      encode(HashRef $data) : Str

    The encode method encodes and returns the data provided.

    encode example #1

        # given: synopsis
      
        $store->encode({ status => 'ok' });

 keys

      keys(Str @keys) : ArrayRef[Str]

    The keys method returns a list of keys under the namespace of the
    datastore or provided key.

    keys example #1

        # given: synopsis
      
        my $keys = $store->keys('zing:main:global:model:temp');

    keys example #2

        # given: synopsis
      
        $store->send('zing:main:global:model:temp', { status => 'ok' });
      
        my $keys = $store->keys('zing:main:global:model:temp');

 lpull

      lpull(Str $key) : Maybe[HashRef]

    The lpull method pops data off of the top of a list in the datastore.

    lpull example #1

        # given: synopsis
      
        $store->lpull('zing:main:global:model:items');

    lpull example #2

        # given: synopsis
      
        $store->rpush('zing:main:global:model:items', { status => 'ok' });
      
        $store->lpull('zing:main:global:model:items');

 lpush

      lpush(Str $key, HashRef $val) : Int

    The lpush method pushed data onto the top of a list in the datastore.

    lpush example #1

        # given: synopsis
      
        $store->lpush('zing:main:global:model:items', { status => '1' });

    lpush example #2

        # given: synopsis
      
        $store->lpush('zing:main:global:model:items', { status => '0' });
      
        $store->lpush('zing:main:global:model:items', { status => '0' });

 recv

      recv(Str $key) : Maybe[HashRef]

    The recv method fetches and returns data from the datastore by its key.

    recv example #1

        # given: synopsis
      
        $store->recv('zing:main:global:model:temp');

    recv example #2

        # given: synopsis
      
        $store->send('zing:main:global:model:temp', { status => 'ok' });
      
        $store->recv('zing:main:global:model:temp');

 rpull

      rpull(Str $key) : Maybe[HashRef]

    The rpull method pops data off of the bottom of a list in the
    datastore.

    rpull example #1

        # given: synopsis
      
        $store->rpull('zing:main:global:model:items');

    rpull example #2

        # given: synopsis
      
        $store->rpush('zing:main:global:model:items', { status => 1 });
        $store->rpush('zing:main:global:model:items', { status => 2 });
      
        $store->rpull('zing:main:global:model:items');

 rpush

      rpush(Str $key, HashRef $val) : Int

    The rpush method pushed data onto the bottom of a list in the
    datastore.

    rpush example #1

        # given: synopsis
      
        $store->rpush('zing:main:global:model:items', { status => 'ok' });

    rpush example #2

        # given: synopsis
      
        $store->rpush('zing:main:global:model:items', { status => 'ok' });
      
        $store->rpush('zing:main:global:model:items', { status => 'ok' });

 send

      send(Str $key, HashRef $val) : Str

    The send method commits data to the datastore with its key and returns
    truthy.

    send example #1

        # given: synopsis
      
        $store->send('zing:main:global:model:temp', { status => 'ok' });

 size

      size(Str $key) : Int

    The size method returns the size of a list in the datastore.

    size example #1

        # given: synopsis
      
        my $size = $store->size('zing:main:global:model:items');

    size example #2

        # given: synopsis
      
        $store->rpush('zing:main:global:model:items', { status => 'ok' });
      
        my $size = $store->size('zing:main:global:model:items');

 slot

      slot(Str $key, Int $pos) : Maybe[HashRef]

    The slot method returns the data from a list in the datastore by its
    index.

    slot example #1

        # given: synopsis
      
        my $model = $store->slot('zing:main:global:model:items', 0);

    slot example #2

        # given: synopsis
      
        $store->rpush('zing:main:global:model:items', { status => 'ok' });
      
        my $model = $store->slot('zing:main:global:model:items', 0);

 test

      test(Str $key) : Int

    The test method returns truthy if the specific key (or datastore)
    exists.

    test example #1

        # given: synopsis
      
        $store->rpush('zing:main:global:model:items', { status => 'ok' });
      
        $store->test('zing:main:global:model:items');

    test example #2

        # given: synopsis
      
        $store->drop('zing:main:global:model:items');
      
        $store->test('zing:main:global:model:items');

AUTHOR

    Al Newkirk, awncorp@cpan.org

LICENSE

    Copyright (C) 2011-2019, Al Newkirk, et al.

    This is free software; you can redistribute it and/or modify it under
    the terms of the The Apache License, Version 2.0, as elucidated in the
    "license file"
    <https://github.com/iamalnewkirk/zing-store-rocks/blob/master/LICENSE>.

PROJECT

    Wiki <https://github.com/iamalnewkirk/zing-store-rocks/wiki>

    Project <https://github.com/iamalnewkirk/zing-store-rocks>

    Initiatives <https://github.com/iamalnewkirk/zing-store-rocks/projects>

    Milestones
    <https://github.com/iamalnewkirk/zing-store-rocks/milestones>

    Contributing
    <https://github.com/iamalnewkirk/zing-store-rocks/blob/master/CONTRIBUTE.md>

    Issues <https://github.com/iamalnewkirk/zing-store-rocks/issues>



CHANGES
Changelog for Zing-Store-Rocks

0.01 2020-12-26T19:19:53
 - Add release updates
 - Initial implementation (#2)
 - Initial commit


lib/Zing/Store/Rocks.pm
package Zing::Store::Rocks;

use 5.014;

use strict;
use warnings;

use registry 'Zing::Types';
use routines;

use Data::Object::Class;
use Data::Object::ClassHas;

extends 'Zing::Store';

use File::Spec;
use RocksDB;

# VERSION

# ATTRIBUTES

has root => (
  is => 'ro',
  isa => 'Str',
  new => 1,
);

fun new_root($self) {
  my $root = File::Spec->catfile(
    File::Spec->curdir, 'zing',
  );
  mkdir $root;
  $root
}

has client => (
  is => 'ro',
  isa => 'InstanceOf["RocksDB"]',
  new => 1,
);

fun new_client($self) {
  RocksDB->new($self->root, {create_if_missing => 1})
}

# BUILDERS

fun new_encoder($self) {
  require Zing::Encoder::Dump; Zing::Encoder::Dump->new;
}

# METHODS

method drop(Str $key) {
  return int(!!$self->client->delete($key));
}

method keys(Key $query) {
  my $results = [];
  my $regexp = $query =~ s/\*/.*/gr;
  my $iterator = $self->client->new_iterator->seek_to_first;
  while (my ($key, $value) = $iterator->each) {
      if ($key =~ /^$regexp$/) {
        push @$results, $key;
      }
  }
  return $results;
}

method lpull(Str $key) {
  if (my $data = $self->recv($key)) {
    my $result = shift @{$data->{list}};
    $self->send($key, $data);
    return $result;
  }
  else {
    return undef;
  }
}

method lpush(Str $key, HashRef $val) {
  if (my $data = $self->recv($key)) {
    my $result = unshift @{$data->{list}}, $val;
    $self->send($key, $data);
    return $result;
  }
  else {
    my $data = {list => []};
    my $result = unshift @{$data->{list}}, $val;
    $self->send($key, $data);
    return $result;
  }
}

method recv(Str $key) {
  my $data = $self->client->get($key);
  return $data ? $self->decode($data) : $data;
}

method rpull(Str $key) {
  if (my $data = $self->recv($key)) {
    my $result = pop @{$data->{list}};
    $self->send($key, $data);
    return $result;
  }
  else {
    return undef;
  }
}

method rpush(Str $key, HashRef $val) {
  if (my $data = $self->recv($key)) {
    my $result = push @{$data->{list}}, $val;
    $self->send($key, $data);
    return $result;
  }
  else {
    my $data = {list => []};
    my $result = push @{$data->{list}}, $val;
    $self->send($key, $data);
    return $result;
  }
}

method send(Str $key, HashRef $val) {
  my $set = $self->encode($val);
  $self->client->put($key, $set);
  return 'OK';
}

method size(Str $key) {
  if (my $data = $self->recv($key)) {
    return scalar(@{$data->{list}});
  }
  else {
    return 0;
  }
}

method slot(Str $key, Int $pos) {
  if (my $data = $self->recv($key)) {
    return $data->{list}[$pos];
  }
  else {
    return undef;
  }
}

method test(Str $key) {
  return $self->recv($key) ? 1 : 0;
}

1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.017.
use strict;
use warnings;

use 5.014;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "RocksDB Storage Abstraction",
  "AUTHOR" => "Al Newkirk <awncorp\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Zing-Store-Rocks",
  "LICENSE" => "apache",
  "MIN_PERL_VERSION" => "5.014",
  "NAME" => "Zing::Store::Rocks",
  "PREREQ_PM" => {
    "Data::Object::Class" => "2.02",
    "Data::Object::ClassHas" => "2.01",
    "RocksDB" => "0.05",
    "Zing" => "0.20",
    "routines" => "0.02",
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Data::Object::Class" => "2.02",
    "Data::Object::ClassHas" => "2.01",
    "RocksDB" => "0.05",
    "Test::Auto" => "0.12",
    "Zing" => "0.20",
    "routines" => "0.02",
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.01",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Data::Object::Class" => "2.02",
  "Data::Object::ClassHas" => "2.01",
  "RocksDB" => "0.05",
  "Test::Auto" => "0.12",
  "Zing" => "0.20",
  "routines" => "0.02",
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
